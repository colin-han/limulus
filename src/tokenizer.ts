import { ArrowNode, CommaNode, CommentNode, DateNode, DateTimeNode, ErrorNode, FloatNode, IdentityNode, IntegerNode, LineBreakNode, ParenthesisCloseNode, ParenthesisOpenNode, QuotationMarker, SpaceNode, StringNode, SymbolNode, Token } from "./tokens";
import { Range } from "./range";

const tokenRegexes = [
    { type: 'COMMENT', regex: '\\/\\/[^\\n]*' },
    { type: 'DATE', regex: '(?<d_year>\\d{4})-(?<d_month>\\d{1,2})-(?<d_day>\\d{1,2})' },
    { type: 'DATETIME', regex: '(?<year>\\d{4})-(?<month>\\d{1,2})-(?<day>\\d{1,2}) (?<hour>\\d{1,2}):(?<minute>\\d{1,2})(?::(?<second>\\d{1,2}))?' },
    { type: 'IDENTITY', regex: '\\w[\\w\\d_]*' },
    { type: 'STRING', regex: '(?<quote>[\'`"])(?:(?:\\\\.)|[^\\\\])*?\\k<quote>' },
    { type: 'INTEGER', regex: '[+-]?\\d[\\d_]*' },
    { type: 'FLOAT', regex: '[+-]?\\d[\\d_]+\\.[\\d_]*|\\.\\d[\\d_]*|[+-]?\\d[\\d_](?:[eE][+-]?[\\d_]+)?' },
    { type: 'NEWLINE', regex: '\\n+' },
    { type: 'SPACE', regex: '[ \\t]+' },
    { type: 'SYMBOL', regex: '[!@#$%^&~?.\\\\]+' },
    { type: 'PARENTHESIS_OPEN', regex: '[\\(\\[\\{]' },
    { type: 'PARENTHESIS_CLOSE', regex: '[\\)\\]\\}]' },
    { type: 'COMMA', regex: ',' },
    { type: 'ARROW', regex: '->' },
]

const regStr = tokenRegexes.map(k => `(?<${k.type}>${k.regex})`).join("|");

class TokeniseContext {
    readonly code: string;
    readonly reg: RegExp;

    row = 1;
    rowStartPos = 0;
    lastPos = 0;
    private nextToken: RegExpExecArray | null = null;

    constructor(code: string, reg: RegExp) {
        this.code = code.replace(/\r\n/g, '\n');
        this.reg = reg;
    }

    getNextToken(): RegExpExecArray | null {
        if (this.nextToken !== null) {
            this.lastPos = this.nextToken.index + this.nextToken[0].length;
        }
        this.nextToken = this.reg.exec(this.code);
        return this.nextToken;
    }

    getRangeTo(pos: number): Range {
        return new Range(
            this.row,
            this.lastPos - this.rowStartPos + 1,
            this.row,
            pos - this.rowStartPos + 1
        );
    }

    getCodeTo(pos: number): string {
        return this.code.slice(this.lastPos, pos);
    }

    getRangeOf(next: RegExpExecArray, multiline: boolean = false) {
        if (multiline) {
            const current = next[0];
            const currentLen = current.length;
            const lines = current.split(/\n/g);
            const startRow = this.row;
            const startColumn = this.lastPos - this.rowStartPos + 1;
            let endRow = this.row;
            let endColumn = startColumn + currentLen;
            if (lines) {
                endRow += lines.length - 1;
                const endLen = lines[lines.length - 1].length;
                endColumn = endLen + 1;
                this.row = endRow;
                this.rowStartPos = next.index + currentLen - endLen;
            }
            return new Range(startRow, startColumn, endRow, endColumn);

        }
        return new Range(this.row, next.index - this.rowStartPos + 1, this.row, next.index + next[0].length - this.rowStartPos + 1);
    }

    getCodeOf(next: RegExpExecArray) {
        return next[0];
    }

    newLine(lineCount: number) {
        this.row += lineCount;
        this.rowStartPos += lineCount;
    }
}

export function* tokenise(code: string): Generator<Token> {
    // /(?<COMMENT>\/\/[^\n]*)|(?<DATE>(?<d_year>\d{4})-(?<d_month>\d{1,2})-(?<d_day>\d{1,2}))|(?<DATETIME>(?<year>\d{4})-(?<month>\d{1,2})-(?<day>\d{1,2}) (?<hour>\d{1,2}):(?<minute>\d{1,2})(?::(?<second>\d{1,2}))?)|(?<IDENTITY>\w[\w\d_]*)|(?<STRING>(['`"])(?:(?:\.)|[^\\])*?\1)|(?<INTEGER>[+-]?\d[\d_]*)|(?<FLOAT>[+-]?\d[\d_]+\.[\d_]*|\.\d[\d_]*|[+-]?\d[\d_](?:[eE][+-]?[\d_]+)?)|(?<NEWLINE>\n+)|(?<SPACE>[ \t]+)|(?<SYMBOL>[!@#$%^&~?.\\]+)|(?<PARENTHESIS_OPEN>[\(\[\{])|(?<PARENTHESIS_CLOSE>[\)\]\}])|(?<COMMA>,)|(?<ARROW>->)/gm
    const reg = new RegExp(regStr, "mg");
    const context = new TokeniseContext(code, reg);

    let next: RegExpExecArray | null;
    while ((next = context.getNextToken()) != null) {
        const current = next[0];

        const currentLen = current.length;

        if (next.index != context.lastPos) {
            if (!next.groups?.SPACE && !next.groups?.NEWLINE) {
                yield new ErrorNode(
                    context.getRangeTo(next.index + currentLen),
                    context.getCodeTo(next.index + currentLen),
                    "Unexpected token"
                );
                continue;
            }

            yield new ErrorNode(
                context.getRangeTo(next.index),
                context.getCodeTo(next.index),
                "Unexpected token"
            );
        }

        if (next.groups?.SPACE) {
            yield new SpaceNode(context.getRangeOf(next), current, currentLen);
        } else if (next.groups?.STRING) {
            yield new StringNode(context.getRangeOf(next, true), current, current[0] as QuotationMarker);
        } else if (next.groups?.INTEGER) {
            yield new IntegerNode(context.getRangeOf(next), current, parseInt(current));
        } else if (next.groups?.FLOAT) {
            yield new FloatNode(context.getRangeOf(next), current, parseFloat(current));
        } else if (next.groups?.COMMENT) {
            yield new CommentNode(context.getRangeOf(next), current);
        } else if (next.groups?.NEWLINE) {
            yield new LineBreakNode(context.getRangeOf(next, true), current, currentLen);
            // context.newLine(currentLen);
        } else if (next.groups?.SYMBOL) {
            yield new SymbolNode(context.getRangeOf(next), current);
        } else if (next.groups?.PARENTHESIS_OPEN) {
            yield new ParenthesisOpenNode(context.getRangeOf(next), current);
        } else if (next.groups?.PARENTHESIS_CLOSE) {
            yield new ParenthesisCloseNode(context.getRangeOf(next), current);
        } else if (next.groups?.COMMA) {
            yield new CommaNode(context.getRangeOf(next));
        } else if (next.groups?.ARROW) {
            yield new ArrowNode(context.getRangeOf(next));
        } else if (next.groups?.DATE) {
            yield new DateNode(context.getRangeOf(next), current, new Date(current));
        } else if (next.groups?.DATETIME) {
            yield new DateTimeNode(context.getRangeOf(next), current, new Date(current));
        } else {
            yield new IdentityNode(context.getRangeOf(next), current);
        }
    }
}

