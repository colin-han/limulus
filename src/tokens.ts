import { Range } from "./range";

export type TokenType =
    | 'COMMENT'
    | 'SPACE'
    | 'PARENTHESIS_OPEN'
    | 'PARENTHESIS_CLOSE'
    | 'COMMA'
    | 'SYMBOL'
    | 'LINEBREAK'
    | 'IDENTITY'
    | 'STRING'
    | 'INTEGER'
    | 'FLOAT'
    | 'DATE'
    | 'DATETIME'
    | 'ARROW'
    | 'ERROR';

export type QuotationMarker = '"' | "'" | "`" | '"""';

export interface Token<TType extends TokenType = TokenType> {
    readonly type: TType;
    readonly range: Range;
    readonly text: string;
}

export abstract class BaseToken<TType extends TokenType = TokenType> implements Token<TType> {
    constructor(
        public readonly type: TType,
        public readonly range: Range,
        public readonly text: string
    ) { }
}

export class ErrorNode extends BaseToken<"ERROR"> {
    constructor(range: Range, text: string, public readonly reason: string) {
        super("ERROR", range, text);
    }
}

export class CommentNode extends BaseToken<"COMMENT"> {
    constructor(range: Range, text: string) {
        super("COMMENT", range, text);
    }
}

export class ParenthesisOpenNode extends BaseToken<"PARENTHESIS_OPEN"> {
    constructor(range: Range, text: string) {
        super("PARENTHESIS_OPEN", range, text);
    }
}

export class ParenthesisCloseNode extends BaseToken<"PARENTHESIS_CLOSE"> {
    constructor(range: Range, text: string) {
        super("PARENTHESIS_CLOSE", range, text);
    }
}

export class CommaNode extends BaseToken<"COMMA"> {
    constructor(range: Range) {
        super("COMMA", range, ",");
    }
}

export class SymbolNode extends BaseToken<"SYMBOL"> {
    constructor(range: Range, text: string) {
        super("SYMBOL", range, text);
    }
}

export class LineBreakNode extends BaseToken<"LINEBREAK"> {
    constructor(range: Range, text: string, public readonly count: number) {
        super("LINEBREAK", range, text);
    }
}

export class SpaceNode extends BaseToken<"SPACE"> {
    constructor(range: Range, text: string, public readonly size: number) {
        super("SPACE", range, text);
    }
}

export class IdentityNode extends BaseToken<"IDENTITY"> {
    constructor(
        range: Range,
        text: string
    ) {
        super("IDENTITY", range, text);
    }
}

export class StringNode extends BaseToken<"STRING"> {
    constructor(
        range: Range,
        text: string,
        public readonly quot: QuotationMarker,
    ) {
        super("STRING", range, text);
    }
}

export abstract class NumberNode<TType extends "INTEGER" | "FLOAT"> extends BaseToken<TType> {
    constructor(
        type: TType,
        range: Range,
        text: string,
        public readonly value: number
    ) {
        super(type, range, text);
    }
}

export class IntegerNode extends NumberNode<"INTEGER"> {
    constructor(range: Range, text: string, value: number) {
        super("INTEGER", range, text, value);
    }
}
export class FloatNode extends NumberNode<"FLOAT"> {
    constructor(range: Range, text: string, value: number) {
        super("FLOAT", range, text, value);
    }
}

export class DateNode extends BaseToken<'DATE'> {
    constructor(range: Range, text: string, public readonly value: Date) {
        super("DATE", range, text);
    }
}

export class DateTimeNode extends BaseToken<'DATETIME'> {
    constructor(range: Range, text: string, public readonly value: Date) {
        super("DATETIME", range, text);
    }
}

export class ArrowNode extends BaseToken<'ARROW'> {
    constructor(range: Range) {
        super("ARROW", range, "->");
    }
}