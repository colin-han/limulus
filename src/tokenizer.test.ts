import { tokenise } from "./tokenizer";
import { StringNode } from "./tokens";

test("should tokenize as expected", () => {
  const doc = `
    name
      test
    `;

  const tokens = [...tokenise(doc)];
  expect(tokens).toMatchSnapshot();
})

test("should correctly to tokenize the string", () => {
  const doc = `"abc
  def"`;

  const tokens = [...tokenise(doc)];
  const token = tokens[0] as StringNode;
  expect(token.type).toBe("STRING");
  expect(token.quot).toBe('"');
  expect(token.range).toMatchObject({ startRow: 1, endRow: 2, startColumn: 1, endColumn: 7 });
})

test("should tokenise correctly string with other token", () => {
  const doc = `otherToken 'abc
  def
  foo bar' token2`;

  const tokens = [...tokenise(doc)];
  const token1 = tokens[0];
  expect(token1.type).toBe("IDENTITY");
  expect(token1.text).toBe("otherToken");
  const token2 = tokens[1];
  expect(token2.type).toBe("SPACE");
  const token3 = tokens[2];
  expect(token3.type).toBe("STRING");
  expect((token3 as StringNode).quot).toBe("'");
  expect(token3.range).toMatchObject({ startRow: 1, startColumn: 12, endRow: 3, endColumn: 11 });
  const token5 = tokens[4];
  expect(token5).toMatchObject({ type: "IDENTITY" });
})

test("should tokenize correctly string with line break", () => {
  const doc = `
  \`test
  multi-line\`
  otherToken`;

  const tokens = [...tokenise(doc)];
  expect(tokens[0]).toMatchObject({ type: "LINEBREAK" });
  expect(tokens[1]).toMatchObject({ type: "SPACE" });
  expect(tokens[2]).toMatchObject({
    type: "STRING",
    quot: '`',
    range: { startRow: 2, startColumn: 3, endRow: 3, endColumn: 14 }
  })
})

test("should parse directive correctly", () => {
  const doc = `@unknown(param1, param2)`;

  const tokens = [...tokenise(doc)];
  expect(tokens).toHaveLength(8);
  expect(tokens[0]).toMatchObject({ type: "SYMBOL", text: "@" });
  expect(tokens[1]).toMatchObject({ type: "IDENTITY", text: "unknown" });
  expect(tokens[2]).toMatchObject({ type: "PARENTHESIS_OPEN", text: "(" });
  expect(tokens[3]).toMatchObject({ type: "IDENTITY", text: "param1" });
  expect(tokens[4]).toMatchObject({ type: "COMMA", text: "," });
  expect(tokens[5]).toMatchObject({ type: "SPACE", text: " " });
  expect(tokens[6]).toMatchObject({ type: "IDENTITY", text: "param2" });
  expect(tokens[7]).toMatchObject({ type: "PARENTHESIS_CLOSE", text: ")" });
})