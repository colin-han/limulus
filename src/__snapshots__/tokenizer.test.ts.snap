// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should tokenize as expected 1`] = `
[
  LineBreakNode {
    "count": 1,
    "range": Range {
      "endColumn": 1,
      "endRow": 2,
      "startColumn": 1,
      "startRow": 1,
    },
    "text": "
",
    "type": "LINEBREAK",
  },
  SpaceNode {
    "range": Range {
      "endColumn": 5,
      "endRow": 2,
      "startColumn": 1,
      "startRow": 2,
    },
    "size": 4,
    "text": "    ",
    "type": "SPACE",
  },
  IdentityNode {
    "range": Range {
      "endColumn": 9,
      "endRow": 2,
      "startColumn": 5,
      "startRow": 2,
    },
    "text": "name",
    "type": "IDENTITY",
  },
  LineBreakNode {
    "count": 1,
    "range": Range {
      "endColumn": 1,
      "endRow": 3,
      "startColumn": 9,
      "startRow": 2,
    },
    "text": "
",
    "type": "LINEBREAK",
  },
  SpaceNode {
    "range": Range {
      "endColumn": 7,
      "endRow": 3,
      "startColumn": 1,
      "startRow": 3,
    },
    "size": 6,
    "text": "      ",
    "type": "SPACE",
  },
  IdentityNode {
    "range": Range {
      "endColumn": 11,
      "endRow": 3,
      "startColumn": 7,
      "startRow": 3,
    },
    "text": "test",
    "type": "IDENTITY",
  },
  LineBreakNode {
    "count": 1,
    "range": Range {
      "endColumn": 1,
      "endRow": 4,
      "startColumn": 11,
      "startRow": 3,
    },
    "text": "
",
    "type": "LINEBREAK",
  },
  SpaceNode {
    "range": Range {
      "endColumn": 5,
      "endRow": 4,
      "startColumn": 1,
      "startRow": 4,
    },
    "size": 4,
    "text": "    ",
    "type": "SPACE",
  },
]
`;
